### ISSUE TITLE
Support for Lake Formation tag inheritance

### ISSUE THREAD

**Issue by dacreify** [dbt-athena/dbt-athena#452]
We apply Lake Formation tags at the database level that are subsequently inherited by tables and columns. Currently `dbt-athena` attempts to remove all tags from tables, even those that are inherited, preventing our Lake Formation usage pattern from working.

The `GetResourceLFTags` API call returns full information sufficient to infer tag heredity for a given resource, so it would be possible for `dbt-athena` to not attempt to remove inherited tags this way. They could still be overridden when present at the table or column level in DBT metadata.

Alternatively a simple list of tags intended for inheritance only and not override could be supplied and ignored during the "remove all" phase of applying tags. This supports our use case where DBT principals don't even have permission to assign the tags in question.

**Comment by svdimchenko** [comment]
@dacreify could you please describe your lakeformation + dbt usage pattern ? I'm just not sure if it's OK to have 2 points to manage lf tags: dbt and terraform for instance. Why don't you use only 1 tool for that ?
In my practice, I use only dbt for tags management and there are no issues with such approach. I'm wondering if having 2 management tools will lead to some mess.

**Comment by dacreify** [comment]
@svdimchenko Sure I can elaborate. We use Terraform to provision our Glue databases for `dbt-athena` as well as DB-level tag associations for them, and we use DBT to manage table and column level tags. There's a particular tag we use for isolating data access into channels that's always applied at the DB level and should never be overridden. The principal that runs the `dbt-athena` project doesn't even have permission to associate this tag so that it's impossible to break the boundary between two channels by either forgetting to assign the tag or assigning it incorrectly. Without the change in the PR we therefore just get `permission denied` type errors from the Lake Formation API as when `dbt-athena` tries to remove the inherited tag.

**Comment by svdimchenko** [comment]
I see. The question is if you put this inherited tag into the models config (it can be done in `dbt_project.yml` generally for all models or set by schema) you don't need to add inherited tags behaviour. And you won't face permissions error as these tags will be set on model's level. Did I miss smth ?

**Comment by nicor88** [comment]
I have a similar setup to @dacreify - We do not want to use dbt to create a database and assign the main lf tags to the dbs, because we simply have too many, and we can simply do with some loops in terraform or cdk.

To work around what @dacreify, our analytics engineers (and dbt IAM role) is allowed to assign table and columns tags.

@svdimchenko but I believe that what is being raised here is really relevant, we cannot force people to use one single approach to handle lf tags.

**Comment by svdimchenko** [comment]
@nicor88 I use this approach as well. I setup lf-tags for databases with terraform and lf-tags for table and columns with dbt.
And I agree with you and @dacreify this feature will bring us more flexibility for lf-tags management. I just was wondering if that will not lead to mess somehow.

**Comment by nicor88** [comment]
>I just was wondering if that will not lead to mess somehow.
@svdimchenko you have a good point, indeed, we need to be sure about the proposed approach - I didn't deep dive yet on what is proposed in the codebase.

**Comment by dacreify** [comment]
> The question is if you put this inherited tag into the models config (it can be done in `dbt_project.yml` generally for all models or set by schema) you don't need to add inherited tags behaviour.

My understanding is that the `lf_tags_config` config is not merged when applied at multiple levels but rather overridden. For example if I set my channel tag at the project level and then want to have further tags defined at the model level the latter will overwrite the former and my project-level tag will be lost. (This is why I made `lf_inherited_tags` a new top-level config instead of nested it inside `lf_tags_config`.)

Also our access control scheme requires that there's no way principals running DBT can override the database-level tag, so the permissions errors are actually as intended.

I tried to make my solution as narrow as possible to just support the case of wanting to inherit database-level tags that are never overridden by DBT metadata and spelled this out in the docs.
